stdlib/universe/kaufmansAMA.go

```diff
@@ -0,0 +1,259 @@
+	N       int64    `json:"n"`
+	Columns []string `json:"columns"`
+}
+
+func newkamaProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	spec, ok := qs.(*KamaOpSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+
+	return &KamaProcedureSpec{
+		N:       spec.N,
+		Columns: spec.Columns,
+	}, nil
+}
+
+func (s *KamaProcedureSpec) Kind() plan.ProcedureKind {
+	return kamaKind
+}
+
+func (s *KamaProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := new(KamaProcedureSpec)
+	*ns = *s
+	return ns
+}
+
+// TriggerSpec implements plan.TriggerAwareProcedureSpec
+func (s *KamaProcedureSpec) TriggerSpec() plan.TriggerSpec {
+	return plan.NarrowTransformationTriggerSpec{}
+}
+
+func createkamaTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	s, ok := spec.(*KamaProcedureSpec)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	d := execute.NewDataset(id, mode, cache)
+	t := NewkamaTransformation(d, cache, s)
+	return t, d, nil
+}
+
+type kamaTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+
+	n       int64
+	columns []string
+}
+
+func NewkamaTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *KamaProcedureSpec) *kamaTransformation {
+	return &kamaTransformation{
+		d:     d,
+		cache: cache,
+
+		n:       spec.N,
+		columns: spec.Columns,
+	}
+}
+
+func (t *kamaTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *kamaTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("KAMA found duplicate table with key: %v", tbl.Key())
+	}
+	cols := tbl.Cols()
+	doKAMA := make([]bool, len(cols))
+	for j, c := range cols {
+		found := false
+		for _, label := range t.columns {
+			if c.Label == label {
+				if c.Type != flux.TInt && c.Type != flux.TUInt && c.Type != flux.TFloat {
+					return fmt.Errorf("cannot take KAMA of column %s (type %s)", c.Label, c.Type.String())
+				}
+				found = true
+				break
+			}
+		}
+
+		if found {
+			mac := c
+			mac.Type = flux.TFloat
+			_, err := builder.AddCol(mac)
+			if err != nil {
+				return err
+			}
+			doKAMA[j] = true
+		} else {
+			_, err := builder.AddCol(c)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 || len(cr.Cols()) == 0 {
+			return nil
+		}
+
+		for j, c := range cr.Cols() {
+			if !doKAMA[j] {
+				for i := int(t.n); i < cr.Len(); i++ {
+					if err := builder.AppendValue(j, execute.ValueForRow(cr, i, j)); err != nil {
+						return err
+					}
+				}
+			} else {
+				var values []float64
+				switch c.Type {
+				case flux.TInt:
+					arr := cr.Ints(j)
```

Comment:
I don't think this will work quite right for large tables.  Tables are partitioned into multiple "blocks" that by default can have up to 1000 rows.  The issue is that `table.Do` will be called multiple times for each block.  So at the 1001st row, it will start over assuming that it's a new table.

EMA uses data structures in a package called `moving_average`

Reply:
EMA uses data structures from the pacakge `flux/internal/moving_average` to make this work right.  Since this is also a moving average, I wonder if you could reuse this code?
https://github.com/influxdata/flux/blob/master/stdlib/universe/exponential_moving_average.go#L154
https://github.com/influxdata/flux/tree/master/internal/moving_average